[PyTorch] Avoid atomic decrement for final reference to intrusive_ptr_target

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117987
* #117986

On 64-bit platforms, we can detect when we're destroying the final strong reference to an intrusive_ptr_target and skip atomic decrement operations in that case. See a similar optimization in libstdc++ shared_ptr (but apparently not in libc++ shared_ptr): https://github.com/gcc-mirror/gcc/commit/dbf8bd3c2f2cd2d27ca4f0fe379bd9490273c6d7

Differential Revision: [D52860300](https://our.internmc.facebook.com/intern/diff/D52860300/)

swolchok committed:
[PyTorch] Avoid atomic decrement for final reference to intrusive_ptr_target

On 64-bit platforms, we can detect when we're destroying the final strong reference to an intrusive_ptr_target and skip atomic decrement operations in that case. See a similar optimization in libstdc++ shared_ptr (but apparently not in libc++ shared_ptr): https://github.com/gcc-mirror/gcc/commit/dbf8bd3c2f2cd2d27ca4f0fe379bd9490273c6d7

Differential Revision: [D52860300](https://our.internmc.facebook.com/intern/diff/D52860300/)

[ghstack-poisoned]

swolchok committed:
use std::memory_order_acquire for refcount load on "[PyTorch] Avoid atomic decrement for final reference to intrusive_ptr_target"


On 64-bit platforms, we can detect when we're destroying the final strong reference to an intrusive_ptr_target and skip atomic decrement operations in that case. See a similar optimization in libstdc++ shared_ptr (but apparently not in libc++ shared_ptr): https://github.com/gcc-mirror/gcc/commit/dbf8bd3c2f2cd2d27ca4f0fe379bd9490273c6d7

Differential Revision: [D52860300](https://our.internmc.facebook.com/intern/diff/D52860300/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Avoid atomic decrement for final reference to intrusive_ptr_target"


On 64-bit platforms, we can detect when we're destroying the final strong reference to an intrusive_ptr_target and skip atomic decrement operations in that case. See a similar optimization in libstdc++ shared_ptr (but apparently not in libc++ shared_ptr): https://github.com/gcc-mirror/gcc/commit/dbf8bd3c2f2cd2d27ca4f0fe379bd9490273c6d7

Differential Revision: [D52860300](https://our.internmc.facebook.com/intern/diff/D52860300/)

[ghstack-poisoned]

malfet commented on c10/util/intrusive_ptr.h:
```diff
@@ -277,27 +278,62 @@ class intrusive_ptr final {
+    }
+  }
+
+  #pragma GCC diagnostic push
+  #pragma GCC diagnostic ignored "-Wstrict-aliasing"
```
There are `C10_DIAGNOSTIC_PUSH_AND_IGNORED_IF_DEFINED`, which probably be more appropriate here
https://github.com/pytorch/pytorch/blob/4a49e2b52d90a0f05fc7ead1aeb3fbeab014a2ad/c10/macros/Macros.h#L540

swolchok committed:
Update on "[PyTorch] Avoid atomic decrement for final reference to intrusive_ptr_target"


On 64-bit platforms, we can detect when we're destroying the final strong reference to an intrusive_ptr_target and skip atomic decrement operations in that case. See a similar optimization in libstdc++ shared_ptr (but apparently not in libc++ shared_ptr): https://github.com/gcc-mirror/gcc/commit/dbf8bd3c2f2cd2d27ca4f0fe379bd9490273c6d7

Differential Revision: [D52860300](https://our.internmc.facebook.com/intern/diff/D52860300/)

[ghstack-poisoned]
